### PR TITLE
Respect texture filtering when importing GLTF

### LINKED ISSUE #45206: [3.2.3] Texture filtering not going off
**Godot version:**
Godot 3.2.3

**OS/device including version:**
Gentoo Linux, Nvidia gtx 850m (nvidia-drivers) gles3
USE flags:
[image](https://user-images.githubusercontent.com/39315312/104717589-d1332e00-5729-11eb-9a92-3a13515c6813.png)

**Issue description:**
Nearest neighbor not applying when filter is disabled on import, textures are always filtered.

**Minimal reproduction project:**
[game.zip](https://github.com/godotengine/godot/files/5820085/game.zip)
Behavior in my system:
[image](https://user-images.githubusercontent.com/39315312/104717798-28d19980-572a-11eb-80cb-166f6efab5df.png)

### LINKED ISSUE #61232: GLB/GLTF import, texture sampling missing on imported materials
### Godot version

4.0 alpha 7

### System information

Win 10

### Issue description

I tried importing a *.glft model from blender with the albedo filtering set to "closest". The resulting mesh has linear-mipmap filtering instead.

In *.glft textures may reference a sampler:
```
    "textures": [
        {
            "sampler": 0,
            "source": 2
        }
    ]
```
Samplers hold filtering and wrapping data:
```
    "samplers": [
        {
            "magFilter": 9729,
            "minFilter": 9987,
            "wrapS": 10497,
            "wrapT": 10497
        }
    ]
```
https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#_filtering

[gltf_texture.h](https://github.com/godotengine/godot/blob/master/modules/gltf/gltf_texture.h) does not mention the sampler, so I assume it is just missing and that the wrapping settings are not imported either.

### Steps to reproduce

Try importing a *.glft with filtering mode `NEAREST`/`LINEAR`. Model is provided in the attachment.
Then check `Mesh/Surface1/Material/Sampling/Filter` or zoom deep into the provided models dead, uncaring eyes and fail to see pixels.

### Minimal reproduction project

[nearest_cube.zip](https://github.com/godotengine/godot/files/8743557/nearest_cube.zip)

### PR DESCRIPTION
**Pull request by LunaticInAHat**
This PR updates the GLTF importer, such that imported materials have the correct texture filter modes set (with respect to the sampling specified in the imported file).

This PR builds on work that was started in PR #54044, and forward-ports it to 4.x. The delta between making this work in 4.x vs 3.x is fairly significant, so also have a 3.x branch available for after a fix gets merged to master (that 3.x branch is based on what @TheTophatDemon started, with some small corrections that make it work properly on the GLTF files they were having trouble importing)

*Bugsquad edit: This closes https://github.com/godotengine/godot/issues/61232.*

### CODE REVIEW

**Comment by LunaticInAHat** [comment]
As a sidenote that perhaps someone more experienced with C++/Godot could help me with: I would liked to have kept the filter mode enum as an enum class, for a little bit of enhanced type safety (as TheTophatDemon originally had it), but when declaring it as `enum class` and trying to use it as a return type / argument type (for e.g. `set_min_filter()`), the build just ends up blowing up with errors such as:
`./core/variant/binder_common.h:705:15: error: no match for 'operator=' (operand types are 'Variant' and 'GLTFTextureSampler::FilterMode')`

I was ultimately able to work around this by just using `int` for the return / argument types instead, and casting to the enum, but this feels like a hack. I'd rather do it the right way, I just don't know how.

**Comment by fire** [comment]
Approved cicd workflows.

**Comment by LunaticInAHat** [comment]
Another side-comment: doctool doesn't seem to mind that I've added a new class (`GLTFTextureSampler`), but no documentation file for that class. Shouldn't it be complaining about that?

**Comment by fire** [comment]
Approved the cicd run. Also note that there is a policy towards squashing prs. There's a pr guide.

https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html

**Comment by LunaticInAHat** [comment]
Can you be more specific about what you would like squashed? Do you want it all compacted down into a single commit? I noted that the guide mentions "we favor commits that bring the codebase from one functional state to another functional state". I tried to make sure that each commit in this PR reflects a functional state (although there is obviously a progression of increasing functionality), but if you'd rather see it mashed into a single commit, I can do that.

**Comment by fire** [comment]
Was busy. I'll review as soon as I can.

Yes. Squash into one commit. You can backup if you want to another branch.

**Comment by LunaticInAHat** [comment]
Squashed as requested.

**Comment by fire** [comment]
I will be using https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/TextureSettingsTest to check.

**Comment by LunaticInAHat** [comment]
I have to admit that I don't think I tested or paid close attention to the wrapping modes when I did this. My usecase was for min/mag filters, so that's what I was looking at. The wrap modes are serialized/deserialized, but at a glance I can't see them actually being applied to the materials anywhere. In fact, as I look at things, I am suddenly filled with uncertainty about whether wrap/mirror is a Material property at all, or instead a property of the underlying Texture.

**Comment by fire** [comment]
I think we should do the filtering modes and punt the wrap mirror modes.

**Comment by LunaticInAHat** [comment]
Okay. Should I remove the logic for serializing/deserializing the wrap modes, or leave it in place, just in case someone gets ambitious in the future and wants to make those modes actually work?

**Comment by fire** [comment]
I'll spend some time reviewing today.

Can you rebase?

**Comment by fire** [comment]
> remove the logic for serializing/deserializing the wrap modes, or leave it in place

I think we should not speculatively add the logic for serialize or deserializing code since we have a policy in other area that comment out code should be deleted. This is not the same, but the idea is the code should not be speculative.

**Comment by fire** [comment]
This appears salvaging.

**Comment by Sleggie** [comment]
This would certainly be an important feature for 3.x

**Comment by LunaticInAHat** [comment]
As I mentioned in the original PR description, I've got a branch that accomplishes something similar for 3.x, but PRs don't get merged when they're adding stuff to 3.x that 4.x doesn't have yet, so there's no point opening a PR for it until this one has been merged.

But, at the same time, I don't feel like I've gotten any meaningful review of this PR, and it doesn't seem that close to being merged. I've rebased it two or three times now, and each time it just languishes for weeks/months -- then by the time anyone gets around to looking at it, there are merge conflicts again, at which point the only feedback is effectively "rebase please?" with no commentary on the actual changes being made, and no indication of whether they are good, bad, or otherwise.

File: modules/gltf/structures/gltf_texture.cpp [hunk 1]
```diff
@@ -33,8 +33,11 @@
 void GLTFTexture::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_src_image"), &GLTFTexture::get_src_image);
 	ClassDB::bind_method(D_METHOD("set_src_image", "src_image"), &GLTFTexture::set_src_image);
+	ClassDB::bind_method(D_METHOD("get_sampler"), &GLTFTexture::get_sampler);
+	ClassDB::bind_method(D_METHOD("set_sampler"), &GLTFTexture::set_sampler);
```
Comment (clayjohn) [inline modules/gltf/structures/gltf_texture.cpp:37]:
```suggestion
	ClassDB::bind_method(D_METHOD("set_sampler", "sampler"), &GLTFTexture::set_sampler);
```

File: modules/gltf/structures/gltf_texture_sampler.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	FilterMode mag_filter = FilterMode::LINEAR;
+	FilterMode min_filter = FilterMode::LINEAR_MIPMAP_LINEAR;
+	WrapMode wrap_s = WrapMode::REPEAT;
+	WrapMode wrap_t = WrapMode::REPEAT;
+};
+
+VARIANT_ENUM_CAST(GLTFTextureSampler::FilterMode);
+VARIANT_ENUM_CAST(GLTFTextureSampler::WrapMode);
+
+#endif
```
Comment (clayjohn) [inline modules/gltf/structures/gltf_texture_sampler.h:163]:
```suggestion
#endif // GLTF_TEXTURE_SAMPLER_H

```

**Review by clayjohn — COMMENTED** [review]
Looks good to me. I've left some suggestions for passing CI. 

I'd like to take the opportunity to apologize for the lack of meaningful review. We are kind of in an awkward place right now with respect to the GLTF module code as we don't have a dedicated maintainer. That means it is up to people who aren't as active in the area and who (like myself) are not as comfortable providing reviews in this area to provide the reviews. The effect of that is PRs slip through the cracks even when they are well done, non-controversial and solving a common problem. So again, my apologies.

**Comment by fire** [comment]
@lyuma Can you take a look?

**Comment by fire** [comment]
@clayjohn If remember correctly, this pr is blocked because of the Godot Engine 4 renderer. It's unclear how to store a data model of something that cannot be rendered in GodotEngine.

1. Godot Engine doesn't support repeat or clamp to edge at all I think.
2. In Godot Engine, the filtering mode is per material rather than per texture.

@clayjohn Can you provide background on these rendering details?

**Comment by LunaticInAHat** [comment]
Thanks for the suggestions on making CI pass, and for the explanation for the delayed review :)

As Fire points out, there is a certain amount of impedance-mismatch between the way that GLTF stores texture samplers, versus how the Godot renderer would like to apply them. I believe that this PR fixes the most common usecase, where users just want nearest filtering on their models (with either wrap+wrap or clamp+clamp), however it can't support all the combinations supported by GLTF. A couple examples of things that won't work are:
 - Materials where the S wrapping mode != the T wrapping mode
 - Materials where different map textures (normal, roughness, etc.) are supposed to be filtered/wrapped differently than the albedo texture
Undoubtedly there are assets out there that use those unsupported combinations.

If I recall correctly, some of those properties used to be per-texture (per-image) in 3.x, so I have assumed that making them global to the material was an intentional choice in 4.x, and that there wouldn't be any interest in changing the material system / renderer to accommodate them.

**Comment by clayjohn** [comment]
For now I would stick within the limits of what is exposed via StandardMaterial/BaseMaterial.

**Comment by LunaticInAHat** [comment]
Yeah, I don't have any intention of expanding this PR to embrace anything beyond what the standard material classes expose. It is unfortunate that the materials don't support all the various sampler combinations that might be exported from art tools, but I don't think this PR should be blocked by that.

As far as anisotropy goes, as far as I can tell, GLTF does not store any information relating to anisotropic filtering (thus it cannot be imported from the GLTF). So, adjusting that will have to be a manual operation performed on the materials in Godot, after import.

It's not 100% clear to me why some of the CI builds are failing, at this point. I think I've solved the issue relating to missing documentation in my latest version of the PR, but the error in "Build godot-cpp test extension" is somewhat opaque to me.

**Comment by fire** [comment]
If you're able to help out, I wanted to do image comparison before and after check using this list of models.

https://github.khronos.org/glTF-Asset-Generator/Output/Positive/Texture_Sampler/

image


I'm busy until the weekend, hope someone can help.

Then, I can approve.

**Comment by LunaticInAHat** [comment]
I already have a 3.x branch prepared, so once this PR gets merged, I'll open the PR for the 3.x backport. Then it should just be a question of testing, and re-review. The 3.x implementation is somewhat different from the 4.x implementation, due to larger changes in how materials & textures work, so testing results from one branch don't necessarily translate to the other.

As far as the image comparison requested by @fire goes, I just want to point out that we expect to fail tests 01 - 04, because they are using wrap modes where S != T.

**Comment by fire** [comment]
Yeah! and if you wish, can save the project and publish it so I can add it as a pr to godot-tests

**Comment by fire** [comment]
I think this is a good enough case for a separate bugfix. uhhh i think we have to add a fake gltf root. But you seem confident

**Review by fire — APPROVED** [review]

File: modules/gltf/structures/gltf_texture_sampler.h [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	enum WrapMode {
+		CLAMP_TO_EDGE = 33071,
+		MIRRORED_REPEAT = 33648,
+		REPEAT = 10497,
+		DEFAULT = REPEAT
+	};
+
+	int get_mag_filter() const {
+		return mag_filter;
+	};
```
Comment (akien-mga) [inline modules/gltf/structures/gltf_texture_sampler.h:59]: Unnecessary semicolon (here and for other methods below).

Comment (LunaticInAHat) [reply modules/gltf/structures/gltf_texture_sampler.h]: Fixed

**Review by akien-mga — APPROVED** [review]

File: modules/gltf/doc_classes/GLTFTextureSampler.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+	<members>
+		<member name="mag_filter" type="int" setter="set_mag_filter" getter="get_mag_filter" default="9729">
+		</member>
+		<member name="min_filter" type="int" setter="set_min_filter" getter="get_min_filter" default="9987">
+		</member>
+		<member name="wrap_s" type="int" setter="set_wrap_s" getter="get_wrap_s" default="10497">
+		</member>
+		<member name="wrap_t" type="int" setter="set_wrap_t" getter="get_wrap_t" default="10497">
+		</member>
+	</members>
```
Comment (clayjohn) [inline modules/gltf/doc_classes/GLTFTextureSampler.xml:18]: These need to be filled it.

**Review by clayjohn — CHANGES_REQUESTED** [review]
Code changes look good now. Once the docs are filled in this should be ready to merge!

File: modules/gltf/doc_classes/GLTFTextureSampler.xml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+<class name="GLTFTextureSampler" inherits="Resource" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<tutorials>
+	</tutorials>
+	<members>
+		<member name="mag_filter" type="int" setter="set_mag_filter" getter="get_mag_filter" default="9729">
+		Texture's magnification filter, used when texture appears larger on screen than the source image.
```
Comment (clayjohn) [inline modules/gltf/doc_classes/GLTFTextureSampler.xml:11]: These need indentation
```suggestion
			Texture's magnification filter, used when texture appears larger on screen than the source image.
```

File: modules/gltf/doc_classes/GLTFTexture.xml [hunk 6]
```diff
@@ -7,6 +7,9 @@
 	<tutorials>
 	</tutorials>
 	<members>
+		<member name="sampler" type="int" setter="set_sampler" getter="get_sampler" default="-1">
+		ID of the texture sampler to use when sampling the image. If -1, then the default texture sampler is used (linear filtering, and repeat wrapping in both axes).
```
Comment (clayjohn) [inline modules/gltf/doc_classes/GLTFTexture.xml:11]:
```suggestion
			ID of the texture sampler to use when sampling the image. If -1, then the default texture sampler is used (linear filtering, and repeat wrapping in both axes).
```

File: modules/gltf/doc_classes/GLTFTextureSampler.xml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+		</member>
+		<member name="min_filter" type="int" setter="set_min_filter" getter="get_min_filter" default="9987">
+		Texture's minification filter, used when the texture appears smaller on screen than the source image.
```
Comment (clayjohn) [inline modules/gltf/doc_classes/GLTFTextureSampler.xml:14]:
```suggestion
			Texture's minification filter, used when the texture appears smaller on screen than the source image.
```

File: modules/gltf/doc_classes/GLTFTextureSampler.xml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+		</member>
+		<member name="wrap_s" type="int" setter="set_wrap_s" getter="get_wrap_s" default="10497">
+		Wrapping mode to use for S-axis (horizontal) texture coordinates.
```
Comment (clayjohn) [inline modules/gltf/doc_classes/GLTFTextureSampler.xml:17]:
```suggestion
			Wrapping mode to use for S-axis (horizontal) texture coordinates.
```

File: modules/gltf/doc_classes/GLTFTextureSampler.xml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+		</member>
+		<member name="wrap_t" type="int" setter="set_wrap_t" getter="get_wrap_t" default="10497">
+		Wrapping mode to use for T-axis (vertical) texture coordinates.
```
Comment (clayjohn) [inline modules/gltf/doc_classes/GLTFTextureSampler.xml:20]:
```suggestion
			Wrapping mode to use for T-axis (vertical) texture coordinates.
```

File: modules/gltf/doc_classes/GLTFTextureSampler.xml [hunk 10]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="GLTFTextureSampler" inherits="Resource" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
```
Comment (clayjohn) [inline modules/gltf/doc_classes/GLTFTextureSampler.xml:6]:
```suggestion
	<brief_description>
		Represents a GLTF texture sampler
	</brief_description>
	<description>
		Represents a texture sampler as defined by the base GLTF spec.
	</description>
```

I have just filled these in based on the current GLTFCamera documentation

**Review by clayjohn — CHANGES_REQUESTED** [review]
CI is failing due to missing indentation. I have made code suggestions to make CI pass and finish filling out the docs.

**Review by clayjohn — APPROVED** [review]

**Comment by clayjohn** [comment]
Thanks for sticking with this one!

**Pull request merged**